src/query/storage/promremote/promremotetest/test_server.go

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2021  Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package promremotetest provides test utilities.
+package promremotetest
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+	"testing"
+
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TestPromServer is a fake http server handling prometheus remote write. Intended for test usage.
+type TestPromServer struct {
+	mu               sync.Mutex
+	lastWriteRequest *prompb.WriteRequest
+	addr             string
+	respErr          error
+	t                *testing.T
+	svr              *http.Server
+}
+
+// NewServer creates new instance of a fake server.
+func NewServer(t *testing.T) *TestPromServer {
+	testPromServer := &TestPromServer{t: t}
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/write", testPromServer.handleWrite)
+
+	testPromServer.svr = &http.Server{Handler: mux}
+	go func() {
+		if err := testPromServer.svr.Serve(listener); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			//nolint: forbidigo
+			fmt.Printf("unexpected testPromServer error %v \n", err)
+		}
+	}()
+	testPromServer.addr = listener.Addr().String()
```

Comment:
nit: You can use `httptest` here which does a few of these things for you:
https://pkg.go.dev/net/http/httptest#NewServer

i.e. 
```go
testPromServer.svr := httptest.NewServer() // creates and starts server
testPromServer.addr = testPromServer.svr.URL

// Later to close
testPromServer.svr.Close()
```